Sources/Yams/Parser.swift

```diff
@@ -108,33 +116,62 @@ public final class Parser {
     /// Constructor.
     public let constructor: Constructor
 
+    /// Encoding
+    public enum Encoding: String {
+        /// Use `YAML_UTF8_ENCODING`
+        case utf8
+        /// Use `YAML_UTF16(BE|LE)_ENCODING`
+        case utf16
+        /// Default encoding can be selected at compile time
+        public static var `default`: Encoding = {
+            let key = "YAMS_ENCODING"
+            if let yamsEncoding = ProcessInfo.processInfo.environment[key],
+                let encoding = Encoding(rawValue: yamsEncoding.lowercased()) {
+                print("`Parser.Encoding.default` is set to `\(encoding)` by `\(key)` environment variable.")
+                return encoding
+            }
+            return "test".utf8.withContiguousStorageIfAvailable({ _ in true }) != nil ? .utf8 : .utf16
+        }()
+    }
+    /// Encoding
+    public let encoding: Encoding
+
     /// Set up Parser.
     ///
     /// - parameter string: YAML string.
     /// - parameter resolver: Resolver, `.default` if omitted.
     /// - parameter constructor: Constructor, `.default` if omitted.
+    /// - parameter encoding: Encoding, `.default` if omitted.
     ///
     /// - throws: `YamlError`.
     public init(yaml string: String,
                 resolver: Resolver = .default,
-                constructor: Constructor = .default) throws {
+                constructor: Constructor = .default,
+                encoding: Encoding = .default) throws {
         yaml = string
         self.resolver = resolver
         self.constructor = constructor
+        self.encoding = encoding
 
         yaml_parser_initialize(&parser)
-#if USE_UTF8
-        yaml_parser_set_encoding(&parser, YAML_UTF8_ENCODING)
-        utf8Slice = string.utf8CString.dropLast()
-        try utf8Slice.withUnsafeBytes(startParse(with:))
-#else
-        // use native endian
-        let isLittleEndian = 1 == 1.littleEndian
-        yaml_parser_set_encoding(&parser, isLittleEndian ? YAML_UTF16LE_ENCODING : YAML_UTF16BE_ENCODING)
-        let encoding: String.Encoding = isLittleEndian ? .utf16LittleEndian : .utf16BigEndian
-        data = yaml.data(using: encoding)!
-        try data.withUnsafeBytes(startParse(with:))
-#endif
+        switch encoding {
+        case .utf8:
+            yaml_parser_set_encoding(&parser, YAML_UTF8_ENCODING)
+            buffer = .utf8View(yaml.utf8)
+            if try yaml.utf8.withContiguousStorageIfAvailable(startParse(with:)) == nil {
```

Comment:
It took me a while to understand what was happening here and why we weren't handling the `else` branch.

A comment or code restructure would be useful to indicate that if the result is non-nil, the operation was successful, and the nil case falls back to converting the string to a UTF8 slice.

Suggestion 1:

```swift
let utf8View = yaml.utf8
buffer = .utf8View(utf8View)
if try utf8View.withContiguousStorageIfAvailable(startParse(with:)) != nil {
    // Attempt to parse with underlying UTF8 String encoding was successful, nothing further to do
} else {
    // Fall back to using UTF8 slice
    let utf8Slice = string.utf8CString.dropLast()
    buffer = .utf8Slice(utf8Slice)
    try utf8Slice.withUnsafeBytes(startParse(with:))
}
```

Or suggestion 2:

```swift
let utf8View = yaml.utf8
buffer = .utf8View(utf8View)
let parsingFailed = try utf8View.withContiguousStorageIfAvailable(startParse(with:)) == nil
guard parsingFailed else { break }

// Attempt to parse with underlying UTF8 String encoding failed.
// Fall back to using UTF8 slice.
let utf8Slice = string.utf8CString.dropLast()
buffer = .utf8Slice(utf8Slice)
try utf8Slice.withUnsafeBytes(startParse(with:))
```

Reply:
Thanks for suggestions! 🙏  I prefer the former.